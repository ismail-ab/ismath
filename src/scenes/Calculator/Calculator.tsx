import * as React from 'react';
import {
  Card,
  CardContent,
  Grid
} from '@material-ui/core';
import {
  Clear,
  Comma,
  Digits,
  Equal,
  Operators,
  Screen
} from './components';
import { compute } from './helpers';
import {
  CalculatorContext,
  ICalculatorContext
} from './contexts';

interface ICalculatorState extends ICalculatorContext { }

class Calculator extends React.Component<{}, ICalculatorState> {
  constructor(props: {}) {
    super(props);

    this.state = {
      concatenatedCharacters: '',
      result: '',
      clearCalculator: this.clearCalculator.bind(this),
      concatenateCharactersOrCompute: this.concatenateCharactersOrCompute.bind(this),
    };
  }

  clearCalculator() {
    this.setState({
      concatenatedCharacters: '',
      result: ''
    });
  }

  concatenateCharactersOrCompute(character: string) {
    this.setState((prevState) => {
      if (character !== '=') {
        return {
          ...prevState,
          concatenatedCharacters: prevState.concatenatedCharacters + character,
        };
      } else {
        return {
          ...prevState,
          result: compute(this.state.concatenatedCharacters)
        };
      }
    });
  }

  render() {
    return (
      <CalculatorContext.Provider value={this.state}>
        <Grid
          container={true}
          direction="column"
          justify="center"
          style={{
            height: '100%',
            width: '100%',
            position: 'absolute'
          }}
        >
          <Grid
            container={true}
            justify="center"
          >
            <Card style={{ width: '348px' }}>
              <CardContent>
                <Screen />
                <Clear />
              </CardContent>
            </Card>
          </Grid>
          <Grid
            container={true}
            direction="row-reverse"
            justify="center"
          >
            <Grid>
              <Operators />
            </Grid>
            <Grid
              container={true}
              justify="center"
              alignItems="center"
              direction="row-reverse"
              spacing={0}
              style={{ width: '260px' }}
            >
              <Digits />
              <Comma />
              <Equal />
            </Grid>
          </Grid>
        </Grid>
      </CalculatorContext.Provider>
    );
  }
}

export default Calculator;
