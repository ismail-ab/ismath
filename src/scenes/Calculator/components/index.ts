export {
  Clear,
  Comma,
  Digits,
  Equal,
  Operators,
  Screen
} from './scenes';