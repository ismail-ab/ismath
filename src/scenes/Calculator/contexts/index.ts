export {
  CalculatorContext,
  ICalculatorContext
} from './CalculatorContext';