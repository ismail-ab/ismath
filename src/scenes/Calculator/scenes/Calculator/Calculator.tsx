import * as React from 'react';
import {
  Card,
  CardContent,
  Grid
} from '@material-ui/core';
import {
  Clear,
  Comma,
  Digits,
  Equal,
  Operators,
  Screen
} from '../components';

interface ICalculatorContainerState {
  concatenatedCharacters: string;
  result: number | string;
}

class Calculator extends React.Component<{}, ICalculatorContainerState> {
  constructor(props: {}) {
    super(props);

    this.state = {
      concatenatedCharacters: '',
      result: ''
    };

    this.clearCalculator = this.clearCalculator.bind(this);
    this.concatenateCharacters = this.concatenateCharacters.bind(this);
    this.computeConcatenatedCharacters = this.computeConcatenatedCharacters.bind(this);
  }

  clearCalculator() {
    this.setState({
      concatenatedCharacters: '',
      result: ''
    });
  }

  concatenateCharacters(character: string) {
    this.setState((prevState) => {
      return {
        ...prevState,
        concatenatedCharacters: prevState.concatenatedCharacters + character,
      };
    });
  }

  computeConcatenatedCharacters() {
    function evaluate(_concatenatedCharacters: string) {
      try {
        return new Function('return ' + _concatenatedCharacters)();
      } catch {
        return 'Expression error';
      }
    }

    this.setState({
      result: evaluate(this.state.concatenatedCharacters)
    });
  }

  render() {
    return (
      <Grid
        container={true}
        direction="column"
        justify="center"
        style={{
          height: '100%',
          width: '100%',
          position: 'absolute',
        }}
      >
        <Grid
          container={true}
          justify="center"
        >
          <Card style={{ width: '348px' }}>
            <CardContent>
              <Screen {...this.state} />
              <Clear clearCalculator={this.clearCalculator} />
            </CardContent>
          </Card>
        </Grid>
        <Grid
          container={true}
          direction="row-reverse"
          justify="center"
        >
          <Grid>
            <Operators concatenateCharacters={this.concatenateCharacters} />
          </Grid>
          <Grid
            container={true}
            justify="center"
            alignItems="center"
            direction="row-reverse"
            spacing={0}
            style={{ width: '260px' }}
          >
            <Digits concatenateCharacters={this.concatenateCharacters} />
            <Comma concatenateCharacters={this.concatenateCharacters} />
            <Equal computeConcatenatedCharacters={this.computeConcatenatedCharacters} />
          </Grid>
        </Grid>
      </Grid>
    );
  }
}

export default Calculator;
