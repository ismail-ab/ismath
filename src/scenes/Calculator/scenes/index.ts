export { default as CalculatorContainer } from './CalculatorContainer';