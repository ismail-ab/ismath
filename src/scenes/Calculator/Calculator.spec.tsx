import { } from 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import Calculator from './Calculator';

describe('Calculator', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Calculator />);

    expect(wrapper).toMatchSnapshot();
  });
});
