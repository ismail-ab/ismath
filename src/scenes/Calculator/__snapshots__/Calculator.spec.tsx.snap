// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<[object Object]
  value={
    Object {
      "clearCalculator": [Function],
      "concatenateCharactersOrCompute": [Function],
      "concatenatedCharacters": "",
      "result": "",
    }
  }
>
  <WithStyles(GridWrapper)
    container={true}
    direction="column"
    justify="center"
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <WithStyles(GridWrapper)
      container={true}
      justify="center"
    >
      <WithStyles(Card)
        style={
          Object {
            "width": "348px",
          }
        }
      >
        <WithStyles(CardContent)>
          <Screen />
          <Clear />
        </WithStyles(CardContent)>
      </WithStyles(Card)>
    </WithStyles(GridWrapper)>
    <WithStyles(GridWrapper)
      container={true}
      direction="row-reverse"
      justify="center"
    >
      <WithStyles(GridWrapper)>
        <Operators />
      </WithStyles(GridWrapper)>
      <WithStyles(GridWrapper)
        alignItems="center"
        container={true}
        direction="row-reverse"
        justify="center"
        spacing={0}
        style={
          Object {
            "width": "260px",
          }
        }
      >
        <Digits />
        <Comma />
        <Equal />
      </WithStyles(GridWrapper)>
    </WithStyles(GridWrapper)>
  </WithStyles(GridWrapper)>
</[object Object]>
`;
